### ISSUE TITLE
program.c, riscv_program_exec(): remove dead code for save/restore of registers

### ISSUE THREAD

**Issue by JanMatCodasip** [riscv/riscv-openocd#982]
The code in riscv_program_exec() dealing with save and restore of registers is not used anymore.

Remove the following:
- 2x for loop in riscv_program_exec() that deal with saving and restoring register values
- saved_registers local array
- writes_xreg array in struct riscv_program

**Comment by en-sc** [comment]
Wouldn't it be better to use the `writes_xreg` array to minimize the possibility of missing a register save like so:
```
+int riscv_program_save_regs(struct riscv_program *p, struct target *t) {
+       for (enum gdb_regno regno = GDB_REGNO_ZERO + 1; regno <= GDB_REGNO_XPR31; ++regno) {
+               if (p->writes_xreg[regno]) {
+                       LOG_TARGET_DEBUG(t, "Saving register %s as used by program", gdb_regno_name(t, regno));
+                       int result = riscv_save_register(t, regno);
+                       if (result != ERROR_OK)
+                               return result;
+               }
+       }
+       return ERROR_OK;
+}
+
 /** Add ebreak and execute the program. */
 int riscv_program_exec(struct riscv_program *p, struct target *t)
 {
 ...
-       riscv_reg_t saved_registers[GDB_REGNO_XPR31 + 1];
-       for (size_t i = GDB_REGNO_ZERO + 1; i <= GDB_REGNO_XPR31; ++i) {
-               if (p->writes_xreg[i]) {
-                       LOG_TARGET_DEBUG(t, "Saving register %d as used by program", (int)i);
-                       int result = riscv_get_register(t, &saved_registers[i], i);
-                       if (result != ERROR_OK)
-                               return result;
-               }
-       }
+       int result = riscv_program_save_regs(p, t);
+       if (result != ERROR_OK)
+               return result;
```
And modifying the users in the following way:
 ```
 int riscv_program_ldr(struct riscv_program *p, enum gdb_regno d, enum gdb_regno b, int offset)
 {
+       p->writes_xreg[d] = true;
        return riscv_program_insert(p, ld(d, b, offset));
 }
 ```

**Comment by en-sc** [comment]
I will prepare a patch with the full suggestion.

**Comment by JanMatCodasip** [comment]
Your suggestion may work. However, be careful about situations where program buffer is executed multiple times before flushing the register cache. Would the second execution of program buffer would cause a dirty value of a register to be saved?

I tend to prefer the current solution: explicit calls to riscv_save_register(). But if you'd prefer the other option, that sounds fine to me, as long as the impact is carefully analyzed.

**Comment by en-sc** [comment]
Calling `riscv_save_register()` on a dirty register does not change the value in the cache, since the register's value is not read from the target -- it is read from the cache (`valid` bit is set).

**Comment by en-sc** [comment]
I've implemented the patch: https://github.com/riscv/riscv-openocd/pull/1003.
However, throughout the codebase `riscv_program_insert()` is used explicitly, which makes automatic tracking of clobbered registers impossible.
Therefore, I believe removing the loop as you suggested will be better.
@JanMatCodasip, can you please implement your suggestion?

**Comment by JanMatCodasip** [comment]
@en-sc Of course. I will submit a pull request shortly (probably on Monday).

**Comment by JanMatCodasip** [comment]
@en-sc I am sorry for delayed response. 

I have just posted the patch with the dead code removal: https://github.com/riscv/riscv-openocd/pull/1012

**Comment by en-sc** [comment]
AFAIU, the issue is fixed by #1012.
